awx/main/tests/functional/api/test_instance_peers.py

```diff
@@ -0,0 +1,331 @@
+            data={"node_type": "execution"},
+            user=admin_user,
+            expect=400,
+        )
+
+    @pytest.mark.parametrize('node_type', ['hop', 'execution'])
+    def test_listener_port_null(self, node_type, admin_user, post):
+        '''
+        listener_port can be None
+        '''
+        post(
+            url=reverse('api:instance_list'),
+            data={"hostname": "abc", "node_type": node_type, "listener_port": None},
+            user=admin_user,
+            expect=201,
+        )
+
+    @pytest.mark.parametrize('node_type, allowed', [('control', False), ('hop', True), ('execution', True)])
+    def test_peers_from_control_nodes_allowed(self, node_type, allowed, post, admin_user):
+        '''
+        only hop and execution nodes can have peers_from_control_nodes set to True
+        '''
+        post(
+            url=reverse('api:instance_list'),
+            data={"hostname": "abc", "peers_from_control_nodes": True, "node_type": node_type, "listener_port": 6789},
+            user=admin_user,
+            expect=201 if allowed else 400,
+        )
+
+    def test_listener_port_is_required(self, admin_user, post):
+        '''
+        if adding instance to peers list, that instance must have listener_port set
+        '''
+        Instance.objects.create(hostname='abc', node_type="hop", listener_port=None)
+        post(
+            url=reverse('api:instance_list'),
+            data={"hostname": "ex", "peers_from_control_nodes": False, "node_type": "execution", "listener_port": None, "peers": ["abc"]},
+            user=admin_user,
+            expect=400,
+        )
+
+    def test_peers_from_control_nodes_listener_port_enabled(self, admin_user, post):
+        '''
+        if peers_from_control_nodes is True, listener_port must an integer
+        Assert that all other combinations are allowed
+        '''
+        Instance.objects.create(hostname='abc', node_type="control")
+        i = 0
+        for node_type in ['hop', 'execution']:
+            for peers_from in [True, False]:
+                for listener_port in [None, 6789]:
+                    # only disallowed case is when peers_from is True and listener port is None
+                    disallowed = peers_from and not listener_port
+                    post(
+                        url=reverse('api:instance_list'),
+                        data={"hostname": f"abc{i}", "peers_from_control_nodes": peers_from, "node_type": node_type, "listener_port": listener_port},
+                        user=admin_user,
+                        expect=400 if disallowed else 201,
+                    )
+                    i += 1
+
+    def test_disallow_modify_peers_control_nodes(self, admin_user, patch):
+        '''
+        for control nodes, peers field should not be
+        modified directly via patch.
+        '''
+        control = Instance.objects.create(hostname='abc', node_type='control')
```

Comment:
pretty much take as given that it would be good to cover hybrid nodes everywhere else in this module unless it explicitly doesn't make sense (but I can't think of circumstances where that would happen).